homeassistant/components/recorder/__init__.py

```diff
@@ -910,33 +914,58 @@ def _process_one_event(self, event):
                 dbevent.event_data = None
             else:
                 dbevent = Events.from_event(event)
-            self.event_session.add(dbevent)
         except (TypeError, ValueError):
             _LOGGER.warning("Event is not JSON serializable: %s", event)
             return
 
+        self.event_session.add(dbevent)
         if event.event_type == EVENT_STATE_CHANGED:
             try:
                 dbstate = States.from_event(event)
-                has_new_state = event.data.get("new_state")
-                if dbstate.entity_id in self._old_states:
-                    old_state = self._old_states.pop(dbstate.entity_id)
-                    if old_state.state_id:
-                        dbstate.old_state_id = old_state.state_id
-                    else:
-                        dbstate.old_state = old_state
-                if not has_new_state:
-                    dbstate.state = None
-                dbstate.event = dbevent
-                self.event_session.add(dbstate)
-                if has_new_state:
-                    self._old_states[dbstate.entity_id] = dbstate
-                    self._pending_expunge.append(dbstate)
-            except (TypeError, ValueError):
+                dbstate_attributes = StateAttributes.from_event(event)
+            except (TypeError, ValueError) as ex:
                 _LOGGER.warning(
-                    "State is not JSON serializable: %s",
+                    "State is not JSON serializable: %s: %s",
                     event.data.get("new_state"),
+                    ex,
                 )
+                return
+
+            dbstate.attributes = None
+            shared_attrs = dbstate_attributes.shared_attrs
+            # Matching attributes found in the pending commit
+            if pending_attributes := self._pending_state_attributes.get(shared_attrs):
+                dbstate.state_attributes = pending_attributes
+            # Matching attributes id found in the cache
+            elif attributes_id := self._state_attributes_ids.get(shared_attrs):
+                dbstate.attributes_id = attributes_id
+            # Matching attributes found in the database
+            elif (
+                attributes := self.event_session.query(StateAttributes.attributes_id)
+                .filter(StateAttributes.hash == dbstate_attributes.hash)
+                .filter(StateAttributes.shared_attrs == shared_attrs)
+                .first()
```

Comment:
What's the expected performance impact on this?
Is the filtering first on hash, then on the attributes to handle hash collisions?

Reply:
Its filtering on the hash and then the attributes if there is a collision.  In most cases it should only have to search one row.  The performance should be negligible.

Reply:
I did some profiling with py-spy and cProfile when I was developing this. It didn't show up in the profile

Reply:
https://maui.koston.org:8123/local/slam_state_machine.svg

Here is one with https://github.com/bdraco/scaletest running.

The fnv1a is really the only thing I could see being optimized (currently being calculated in python), but its still paling in comparison to the other operations.

Reply:
If we ever need to care about the fnv1a speed https://pypi.org/project/fnvhash-c/ or pyhash might be an option.

Reply:
So far no collisions on any of my instances

 select hash,count(hash) as count from state_attributes group by hash having count > 1;

Reply:
Reduced the change we need to hash the attributes here https://github.com/home-assistant/core/pull/68395